include/counter.h

```diff
@@ -22,35 +23,106 @@
 #include <zephyr/types.h>
 #include <stddef.h>
 #include <device.h>
+#include <stdbool.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef void (*counter_callback_t)(struct device *dev, void *user_data);
+struct counter_alarm_cfg;
+
+/** @brief Alarm callback
+ *
+ * @param dev     Pointer to the device structure for the driver instance.
+ * @param cfg     Original structure owning this handler.
+ * @param ticks   Counter value that triggered the callback.
+ */
+typedef void (*counter_alarm_callback_t)(struct device *dev,
+					 const struct counter_alarm_cfg *cfg,
+					 u32_t ticks);
+
+/** @brief Alarm callback structure.
+ *
+ * Used to set an alarm. *
+ * Beware such structure should not be allocated on stack.
+ */
+struct counter_alarm_cfg {
+	u8_t channel_id; /*!< Channel ID. Number of channels is driver
+			  * dependent.
+			  */
+
+	bool absolute; /*!< If true ticks are treated as absolute value, else
+			*   it is relative to the counter reading performed
+			*   during the call.
+			*/
+
+	counter_alarm_callback_t handler; /*!< handler called on alarm
+					   *   (cannot be NULL).
+					   */
+
+	u32_t ticks; /*!< In case of absolute flag is set, maximal value that
+		      *   can be set equals wrap value (counter_get_wrap).
+		      *   Otherwise counter_get_max_relative_alarm() returns
+		      *   maximal value that can be set. If counter is clock
+		      *   driven then ticks can be converted to microseconds
+		      *   (see @ref counter_ticks_to_us). Alternatively,
+		      *   counter implementation may count asynchronous events.
+		      */
+};
+
+typedef void (*counter_wrap_callback_t)(struct device *dev, void *user_data);
```

Comment:
Some description for this would be nice.

Reply:
added